File: oracle/src/aicad_oracle/v1/precheck.py
"""Raw pre-checks of SPEC-v1 §0.5 rule 4 step 3 ([W0-1]): the rejections that must carry a
code although the typed parse (JSON Schema) would fail on them. They run on the JSON value
before schema validation, exactly like `forge_ir::v1::precheck` (the oracle must run the same
checks before its schema validation, SPEC §11 note W7).

| Where | Code |
|---|---|
| any `null` | parse error (no code) |
| `parts[*].features[*].type` unknown | `UNSUPPORTED_FEATURE` |
| `...features[*].v` not a defined version (`1.0`, `0`, `"1"`) | `UNSUPPORTED_FEATURE_VERSION` |
| a parameter's unknown `unit`, missing `value`, present `measure` | `PARAM_INVALID` |
| a Ref (object with `kind` and `q`) whose `card` is not one/some/any/integer ≥ 1 | `INVALID_CARDINALITY` |
| a hole `size` not in `HOLE_SIZES` | `HOLE_SIZE_UNKNOWN` |
| `driving` / `value` on a constraint that is not a dimension | `SKETCH_NOT_A_DIMENSION` |
"""

from __future__ import annotations

from typing import Any

from . import consts
from .ids import is_ref, shown

_U32_MAX = 2**32 - 1


class Problem:
    """A coded rejection: `{code, path}` is normative, `message` and `details` are not."""

    __slots__ = ("code", "path", "message", "details")

    def __init__(self, code: str, path: str, message: str, details: dict | None = None):
        self.code = code
        self.path = path
        self.message = message
        self.details = details if details is not None else {}

    def as_dict(self) -> dict:
        return {"code": self.code, "path": self.path, "message": self.message, "details": self.details}

    def __repr__(self) -> str:
        return f"{self.code} at {self.path}"


def _esc(k: str) -> str:
    return k.replace("~", "~0").replace("/", "~1")


def find_null(v: Any, path: str = "") -> str | None:
    if v is None:
        return path or "/"
    if isinstance(v, list):
        for i, x in enumerate(v):
            p = find_null(x, f"{path}/{i}")
            if p is not None:
                return p
    elif isinstance(v, dict):
        for k, x in v.items():
            p = find_null(x, f"{path}/{_esc(k)}")
            if p is not None:
                return p
    return None


def _is_uint(v: Any) -> bool:
    return isinstance(v, int) and not isinstance(v, bool) and v >= 0


def _shown_value(v: Any) -> str:
    if isinstance(v, bool):
        return "true" if v else "false"
    if isinstance(v, (int, float)):
        return repr(v)
    if isinstance(v, str) and is_ref(v):
        return repr(v)
    if v is None:
        return "null"
    return "<invalid value>"


def _param(p: Any, pp: str, errs: list[Problem]) -> None:
    if not isinstance(p, dict):
        return
    name = shown(p.get("name")) if isinstance(p.get("name"), str) else "<invalid id>"
    allowed = list(consts.param_units())

    def invalid(path: str, reason: str, msg: str) -> None:
        errs.append(Problem("PARAM_INVALID", path, msg, {"name": name, "reason": reason, "allowed": allowed}))

    if "unit" in p:
        u = p["unit"]
        if not (isinstance(u, str) and u in allowed):
            invalid(f"{pp}/unit", "bad-unit", f"unit {_shown_value(u)} is not one of {', '.join(allowed)}")
    else:
        invalid(pp, "bad-unit", "unit is required")
    if "measure" in p:
        invalid(f"{pp}/measure", "measure-deferred",
                "measured parameters are deferred to IR v1.1 (ADR 0013 decision 5)")
    elif "value" not in p:
        invalid(pp, "value-required", "value is required")


def _feature(f: Any, fp: str, errs: list[Problem]) -> None:
    if not isinstance(f, dict):
        return
    ty = f.get("type")
    types = consts.feature_types()
    if not (isinstance(ty, str) and ty in types):
        errs.append(Problem("UNSUPPORTED_FEATURE", f"{fp}/type",
                            f"unknown feature type {_shown_value(ty)}",
                            {"type": _shown_value(ty), "supported": list(types)}))
        return
    if "v" in f:
        v = f["v"]
        supported = consts.feature_versions().get(ty, [])
        if not (_is_uint(v) and v in supported):
            errs.append(Problem("UNSUPPORTED_FEATURE_VERSION", f"{fp}/v",
                                f"{ty} v{_shown_value(v)} is not implemented (supported: {supported})",
                                {"type": ty, "v": _shown_value(v), "supported": supported}))
    if ty == "hole" and "size" in f:
        size = f["size"]
        sizes = list(consts.hole_sizes())
        if not (isinstance(size, str) and size in sizes):
            errs.append(Problem("HOLE_SIZE_UNKNOWN", f"{fp}/size",
                                f"unknown hole size {_shown_value(size)}; use one of {', '.join(sizes)}",
                                {"field": "size", "allowed": sizes}))
    if ty == "sketch" and isinstance(f.get("constraints"), list):
        ctypes, dims = consts.constraint_types(), consts.dimension_types()
        for k, c in enumerate(f["constraints"]):
            if not isinstance(c, dict):
                continue
            t = c.get("type")
            if not isinstance(t, str) or t not in ctypes or t in dims:
                continue
            for key in ("driving", "value"):
                if key in c:
                    cid = shown(c.get("id")) if isinstance(c.get("id"), str) else "<invalid id>"
                    errs.append(Problem("SKETCH_NOT_A_DIMENSION", f"{fp}/constraints/{k}/{key}",
                                        f"`{cid}`: only dimensions (distance, angle, radius, diameter) take {key}",
                                        {"id": cid}))


def _refs(v: Any, path: str, errs: list[Problem]) -> None:
    if isinstance(v, list):
        for i, x in enumerate(v):
            _refs(x, f"{path}/{i}", errs)
    elif isinstance(v, dict):
        if "kind" in v and "q" in v and "card" in v:
            card = v["card"]
            ok = (isinstance(card, str) and card in ("one", "some", "any")) or (
                _is_uint(card) and 1 <= card <= _U32_MAX)
            if not ok:
                errs.append(Problem("INVALID_CARDINALITY", f"{path}/card",
                                    f"card {_shown_value(card)} is not one, some, any or an integer >= 1",
                                    {"field": path, "allowed": ["one", "some", "any", ">= 1"]}))
        for k, x in v.items():
            if k == "capture":
                continue
            _refs(x, f"{path}/{_esc(k)}", errs)


def precheck(v: Any) -> tuple[str | None, list[Problem]]:
    """Returns (parse error message or None, coded problems)."""
    p = find_null(v)
    if p is not None:
        return f"null is not allowed at {p} (omit the field instead)", []
    errs: list[Problem] = []
    if not isinstance(v, dict):
        return None, errs
    if isinstance(v.get("params"), list):
        for i, prm in enumerate(v["params"]):
            _param(prm, f"/params/{i}", errs)
    if isinstance(v.get("parts"), list):
        for pi, part in enumerate(v["parts"]):
            if not isinstance(part, dict):
                continue
            pp = f"/parts/{pi}"
            if isinstance(part.get("params"), list):
                for i, prm in enumerate(part["params"]):
                    _param(prm, f"{pp}/params/{i}", errs)
            if not isinstance(part.get("features"), list):
                continue
            for fi, f in enumerate(part["features"]):
                _feature(f, f"{pp}/features/{fi}", errs)
    _refs(v, "", errs)
    return None, errs

File: forge/crates/forge-ir/src/v1/precheck.rs
//! Raw pre-checks (SPEC-v1 §0.5, [W0-1]): the rejections that must carry a code although the
//! typed parse (serde / JSON Schema) would fail on them. They run on the JSON value before the
//! typed parse, in document order:
//!
//! | Where | Code |
//! |---|---|
//! | any `null` | parse error (IR v1 has no nullable field) |
//! | `parts[*].features[*].type` unknown | `UNSUPPORTED_FEATURE` |
//! | `…features[*].v` not a positive integer, or not a defined version | `UNSUPPORTED_FEATURE_VERSION` |
//! | a parameter's `unit` unknown, `value` missing, `measure` present (deferred to v1.1) | `PARAM_INVALID` |
//! | any Ref (an object with `kind` and `q`) whose `card` is not `one`/`some`/`any`/integer ≥ 1 | `INVALID_CARDINALITY` |
//! | a hole's `size` not in `HOLE_SIZES` | `HOLE_SIZE_UNKNOWN` |
//! | `driving` or `value` on a constraint that is not a dimension | `SKETCH_NOT_A_DIMENSION` |
//!
//! The oracle (W7) and CadScript (W8) must run the same checks before their schema validation.

use serde_json::{Value, json};

use super::LoadError;
use super::features::{FEATURE_TYPES, HoleSize, defined_versions};
use super::ids;
use super::params::ParamUnit;
use super::sketch::{CONSTRAINT_TYPES, DIMENSION_TYPES};
use super::validate::ValidationError;

pub(crate) struct Precheck {
    pub(crate) parse: Option<LoadError>,
    pub(crate) errors: Vec<ValidationError>,
}

pub(crate) fn precheck(v: &Value) -> Precheck {
    let mut out = Precheck {
        parse: None,
        errors: Vec::new(),
    };
    if let Some(p) = find_null(v, "") {
        out.parse = Some(LoadError::Parse {
            message: format!("null is not allowed at {p} (omit the field instead)"),
            path: Some(p),
        });
        return out;
    }
    let errs = &mut out.errors;
    if let Some(params) = v.get("params").and_then(Value::as_array) {
        for (i, p) in params.iter().enumerate() {
            param(p, &format!("/params/{i}"), errs);
        }
    }
    if let Some(parts) = v.get("parts").and_then(Value::as_array) {
        for (pi, part) in parts.iter().enumerate() {
            let pp = format!("/parts/{pi}");
            if let Some(params) = part.get("params").and_then(Value::as_array) {
                for (i, p) in params.iter().enumerate() {
                    param(p, &format!("{pp}/params/{i}"), errs);
                }
            }
            let Some(features) = part.get("features").and_then(Value::as_array) else {
                continue;
            };
            for (fi, f) in features.iter().enumerate() {
                feature(f, &format!("{pp}/features/{fi}"), errs);
            }
        }
    }
    refs(v, "", errs);
    out
}

fn find_null(v: &Value, path: &str) -> Option<String> {
    match v {
        Value::Null => Some(if path.is_empty() {
            "/".into()
        } else {
            path.into()
        }),
        Value::Array(a) => a
            .iter()
            .enumerate()
            .find_map(|(i, x)| find_null(x, &format!("{path}/{i}"))),
        Value::Object(o) => o
            .iter()
            .find_map(|(k, x)| find_null(x, &format!("{path}/{}", escape(k)))),
        _ => None,
    }
}

/// JSON-pointer escaping of one path segment.
fn escape(k: &str) -> String {
    k.replace('~', "~0").replace('/', "~1")
}

fn param(p: &Value, pp: &str, errs: &mut Vec<ValidationError>) {
    let Some(o) = p.as_object() else { return };
    let name = ids::shown(o.get("name").and_then(Value::as_str).unwrap_or(""));
    let invalid = |errs: &mut Vec<ValidationError>, path: String, reason: &str, msg: String| {
        errs.push(ValidationError::new(
            "PARAM_INVALID",
            path,
            msg,
            json!({ "name": name, "reason": reason, "allowed": ParamUnit::ALL }),
        ));
    };
    match o.get("unit") {
        Some(Value::String(u)) if ParamUnit::ALL.contains(&u.as_str()) => {}
        Some(u) => invalid(
            errs,
            format!("{pp}/unit"),
            "bad-unit",
            format!(
                "unit {} is not one of mm, deg, ratio, count, bool",
                shown_value(u)
            ),
        ),
        None => invalid(errs, pp.to_string(), "bad-unit", "unit is required".into()),
    }
    if o.contains_key("measure") {
        invalid(
            errs,
            format!("{pp}/measure"),
            "measure-deferred",
            "measured parameters are deferred to IR v1.1 (ADR 0013 decision 5)".into(),
        );
    } else if !o.contains_key("value") {
        invalid(
            errs,
            pp.to_string(),
            "value-required",
            "value is required".into(),
        );
    }
}

fn feature(f: &Value, fp: &str, errs: &mut Vec<ValidationError>) {
    let Some(o) = f.as_object() else { return };
    let ty = o.get("type").and_then(Value::as_str);
    match ty {
        Some(t) if FEATURE_TYPES.contains(&t) => {}
        _ => {
            errs.push(ValidationError::new(
                "UNSUPPORTED_FEATURE",
                format!("{fp}/type"),
                format!("unknown feature type {}", shown_value(o.get("type").unwrap_or(&Value::Null))),
                json!({ "type": shown_value(o.get("type").unwrap_or(&Value::Null)), "supported": FEATURE_TYPES }),
            ));
            return;
        }
    }
    let ty = ty.unwrap_or_default();
    if let Some(v) = o.get("v") {
        let supported = defined_versions(ty);
        let ok = v
            .as_u64()
            .is_some_and(|n| supported.iter().any(|s| u64::from(*s) == n));
        if !ok {
            errs.push(ValidationError::new(
                "UNSUPPORTED_FEATURE_VERSION",
                format!("{fp}/v"),
                format!(
                    "{ty} v{} is not implemented (supported: {supported:?})",
                    shown_value(v)
                ),
                json!({ "type": ty, "v": shown_value(v), "supported": supported }),
            ));
        }
    }
    if ty == "hole"
        && let Some(size) = o.get("size")
        && size.as_str().is_none_or(|s| HoleSize::parse(s).is_none())
    {
        let allowed: Vec<&str> = HoleSize::ALL.iter().map(|s| s.as_str()).collect();
        errs.push(ValidationError::new(
            "HOLE_SIZE_UNKNOWN",
            format!("{fp}/size"),
            format!(
                "unknown hole size {}; use one of {}",
                shown_value(size),
                allowed.join(", ")
            ),
            json!({ "field": "size", "allowed": allowed }),
        ));
    }
    if ty == "sketch"
        && let Some(cons) = o.get("constraints").and_then(Value::as_array)
    {
        for (k, c) in cons.iter().enumerate() {
            let Some(c) = c.as_object() else { continue };
            let Some(t) = c.get("type").and_then(Value::as_str) else {
                continue;
            };
            if !CONSTRAINT_TYPES.contains(&t) || DIMENSION_TYPES.contains(&t) {
                continue;
            }
            for key in ["driving", "value"] {
                if c.contains_key(key) {
                    let id = ids::shown(c.get("id").and_then(Value::as_str).unwrap_or(""));
                    errs.push(ValidationError::new(
                        "SKETCH_NOT_A_DIMENSION",
                        format!("{fp}/constraints/{k}/{key}"),
                        format!(
                            "`{id}`: only dimensions (distance, angle, radius, diameter) take {key}"
                        ),
                        json!({ "id": id }),
                    ));
                }
            }
        }
    }
}

/// A raw JSON value as messages may show it ([W0-12]): numbers and booleans as they are, strings
/// only when they are well-formed ids, anything else as a placeholder.
fn shown_value(v: &Value) -> String {
    match v {
        Value::Number(_) | Value::Bool(_) | Value::Null => v.to_string(),
        Value::String(s) if ids::is_ref(s) => format!("{s:?}"),
        _ => "<invalid value>".to_string(),
    }
}

/// Every Ref in the document (an object with both `kind` and `q`): check `card`.
fn refs(v: &Value, path: &str, errs: &mut Vec<ValidationError>) {
    match v {
        Value::Array(a) => {
            for (i, x) in a.iter().enumerate() {
                refs(x, &format!("{path}/{i}"), errs);
            }
        }
        Value::Object(o) => {
            if o.contains_key("kind")
                && o.contains_key("q")
                && let Some(card) = o.get("card")
            {
                let ok = match card {
                    Value::String(s) => ["one", "some", "any"].contains(&s.as_str()),
                    Value::Number(n) => n
                        .as_u64()
                        .is_some_and(|n| (1..=u64::from(u32::MAX)).contains(&n)),
                    _ => false,
                };
                if !ok {
                    errs.push(ValidationError::new(
                        "INVALID_CARDINALITY",
                        format!("{path}/card"),
                        format!("card {card} is not one, some, any or an integer >= 1"),
                        json!({ "field": path, "allowed": ["one", "some", "any", ">= 1"] }),
                    ));
                }
            }
            for (k, x) in o {
                if k == "capture" {
                    continue;
                }
                refs(x, &format!("{path}/{}", escape(k)), errs);
            }
        }
        _ => {}
    }
}
